loudml/opentsdb.py

```diff
@@ -0,0 +1,364 @@
+            requests.exceptions.RequestException,
+            requests.exceptions.ConnectionError
+        ) as exn:
+            raise errors.BucketError(self.name, str(exn))
+    return wrapper
+
+
+class OpenTSDBResult(object):
+    """
+    Helper class to parse query result
+    """
+
+    def __init__(self, response):
+        self._response = response
+
+    def __repr__(self):
+        return "OpenTSDB results: {}...".format(str(self._response)[:200])
+
+    def get_points(self):
+        if not self._response:
+            return []
+
+        return self._response[0]['dps'].items()
+
+
+class OpenTSDBClient(object):
+    """
+    Client for OpenTSDB database
+    """
+
+    def __init__(self, host="localhost", port=4242, ssl=False):
+        """
+        Set proper schema based on SSL param
+        """
+        schema = "http"
+        if ssl:
+            schema = "https"
+
+        self.url = "%s://%s:%d" % (schema, host, port)
+
+    def query(self, queries):
+        """
+        Run a list of queries against OpenTSDB
+        """
+        if not isinstance(queries, list):
+            queries = [queries]
+
+        results = []
+
+        for q in queries:
+            time_pred = _build_time_predicates(q["start"], q["end"])
+
+            query_url = "{}/api/query?{}&m={}:{}:{}{}".format(
+                self.url,
+                time_pred,
+                AGGREGATORS.get(q["aggregator"]),
+                q["down_sampler"],
+                q["metric_name"],
+                str(q["tags"])
+            )
+            # TODO: OpenTSDB is capable of running multiple subquries in
+            # one shot. Refactor it to use one request to server
+            resp = requests.get(query_url)
+
+            if resp.status_code >= 200 and resp.status_code < 400:
+                results.append(OpenTSDBResult(resp.json()))
+            elif "No such name for" in resp.text:
+                # Known OpenTSDB issue with 400 and trace if metric not found
+                # https://github.com/OpenTSDB/opentsdb/issues/792
+                results.append(OpenTSDBResult([]))
+            else:
+                logging.error('OpenTSDB error',
+                    resp.status_code, resp.text[:200])
+
+        return results
+
+    def put(self, entry):
+        """
+        Store a point in database
+        """
+        url = "%s/api/put" % self.url
+
+        if entry["sync"]:
+            url = "{}?sync".format(url)
+
+        del entry["sync"]
+
+        resp = requests.post(url, json=entry) # data=entry ?
+
+        if resp.status_code >= 200 and resp.status_code < 400:
+            pass # everything is ok, should we indicate it somehow?
+        else:
+            logging.error('OpenTSDB error', resp.status_code, resp.text[:200])
+
+
+class OpenTSDBBucket(Bucket):
+    """
+    OpenTSDB bucket
+    """
+
+    SCHEMA = Bucket.SCHEMA.extend({
+        Required('addr'): str,
+        Optional('use_ssl', default=False): Boolean(),
```

Comment:
Useful options to add:

- Username and password for basic auth if supported by OpenTSDB
- A `verify_ssl` flag, to verify the server's SSL certificate. Useful if the server is using self-signed certificates.
- Optional for this first pull-request: client certificates parameters

Reply:
Agree